plugin/storage/factory.go

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"flag"
+	"fmt"
+	"os"
+
+	"github.com/spf13/viper"
+	"github.com/uber/jaeger-lib/metrics"
+	"go.uber.org/zap"
+
+	"github.com/jaegertracing/jaeger/plugin"
+	"github.com/jaegertracing/jaeger/plugin/storage/cassandra"
+	"github.com/jaegertracing/jaeger/plugin/storage/es"
+	"github.com/jaegertracing/jaeger/plugin/storage/memory"
+	"github.com/jaegertracing/jaeger/storage"
+	"github.com/jaegertracing/jaeger/storage/dependencystore"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+)
+
+const (
+	// SpanStorageEnvVar is the name of the env var that defines the type of backend used for span storage.
+	SpanStorageEnvVar       = "SPAN_STORAGE"
+	dependencyStorageEnvVar = "DEPENDENCY_STORAGE"
+
+	cassandraStorageType     = "cassandra"
+	elasticsearchStorageType = "elasticsearch"
+	memoryStorageType        = "memory"
+)
+
+var allStorageTypes = []string{cassandraStorageType, elasticsearchStorageType, memoryStorageType}
+
+// Factory implements storage.Factory interface as a meta-factory for storage components.
+// It reads the desired types of storage backends from SPAN_STORAGE and DEPENDENCY_STORAGE
+// environment variable. Allowed values:
+//   * `cassandra` - built-in
+//   * `elasticsearch` - built-in
+//   * `memory` - built-in
+//   * `plugin` - loads a dynamic plugin that implements storage.Factory interface (not supported at the moment)
+type Factory struct {
+	spanStoreType string
+	depStoreType  string
+
+	factories map[string]storage.Factory
+}
+
+// NewFactory creates the meta-factory.
+func NewFactory() (*Factory, error) {
+	f := &Factory{}
+	f.spanStoreType = os.Getenv(SpanStorageEnvVar)
+	if f.spanStoreType == "" {
+		f.spanStoreType = cassandraStorageType
+	}
+	f.depStoreType = os.Getenv(dependencyStorageEnvVar)
+	if f.depStoreType == "" {
+		f.depStoreType = f.spanStoreType
+	}
+	uniqueTypes := map[string]struct{}{
```

Comment:
what black magic is this?

https://play.golang.org/p/nlo0NL_M0M2

Reply:
your example fails with explicit literals, but `uq` works correctly.

Reply:
I don't understand the why though, at runtime uq is also just literals no? so uq should fail too?

Reply:
Well, it doesn't fail if you remove uq2 from the example. If I were to guess, it's due to compiler optimization. When you init the map with the actual literals, compiler can check for uniqueness, and it's a reasonable behavior to fail since such statement is more likely to be a typo. But when you init with dynamic values, compiler cannot check anything, and the initialization is reduced to a sequence of `m[k] = v`.